File: packages/money/src/money.ts
/**
 * Money — the one implementation, shared by server, browser and desktop.
 *
 * This package is the reason the API is TypeScript. The total a resident sees in the
 * app, the total an accountant edits in the console, and the total filed for GST are
 * all produced by this code. There is no second implementation to drift.
 *
 * ## Why not `number`
 *
 * JavaScript `Number` is a binary float. `0.1 + 0.2 === 0.30000000000000004`. Sharing a
 * calculator built on floats would make server and browser wrong *identically*, which is
 * worse than disagreeing — nothing would ever flag it. So money is `Decimal` throughout,
 * and `Money` is a branded type so a bare `number` cannot be passed by accident.
 *
 * ## Rounding
 *
 * Half-up at two decimals for anything invoiced. Indian statutory invoicing expects
 * half-up; banker's rounding would turn ₹0.125 into ₹0.12 and a committee checking
 * against their spreadsheet would report it as a bug. Intermediate values keep four
 * decimals — matching `numeric(18,4)` in Postgres — and only the final amount is
 * quantised.
 */

import Decimal from "decimal.js";

// Half-up, and enough precision that per-sq-ft maths on a 5,000-unit society never
// silently truncates.
Decimal.set({ precision: 34, rounding: Decimal.ROUND_HALF_UP });

/** Branded so a raw `number` is a compile error, not a runtime surprise. */
export type Money = Decimal & { readonly __brand: "Money" };

export const CURRENCY = "INR" as const;

/** Storage precision — matches `numeric(18,4)`. */
const STORAGE_DP = 4;
/** Presentation and invoicing precision — rupees and paise. */
const MONEY_DP = 2;

export const ZERO = new Decimal(0) as Money;

export class MoneyError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "MoneyError";
  }
}

/**
 * Build a money value.
 *
 * `number` is rejected deliberately. Accepting it would reintroduce exactly the
 * floating-point error this module exists to prevent — by the time a float reaches
 * here the precision is already lost, so validating it later cannot help.
 */
export function money(value: string | Decimal): Money {
  if (typeof value === "number") {
    throw new MoneyError(
      "number is not accepted for money — pass a string or Decimal " +
        '(e.g. money("1250.50"), not money(1250.50)).',
    );
  }
  try {
    return new Decimal(value).toDecimalPlaces(
      STORAGE_DP,
      Decimal.ROUND_HALF_UP,
    ) as Money;
  } catch {
    throw new MoneyError(`Not a valid money value: ${String(value)}`);
  }
}

/** Round to rupees and paise, half-up. Use for anything presented or invoiced. */
export function quantise(value: Money | Decimal): Money {
  return new Decimal(value).toDecimalPlaces(
    MONEY_DP,
    Decimal.ROUND_HALF_UP,
  ) as Money;
}

export function add(...values: Money[]): Money {
  return values.reduce<Money>((acc, v) => acc.plus(v) as Money, ZERO);
}

export function subtract(a: Money, b: Money): Money {
  return a.minus(b) as Money;
}

export function multiply(a: Money, factor: string | Decimal): Money {
  return a.times(new Decimal(factor)) as Money;
}

export function isZero(a: Money): boolean {
  return a.isZero();
}

export function isNegative(a: Money): boolean {
  return a.isNegative();
}

export function compare(a: Money, b: Money): -1 | 0 | 1 {
  return a.comparedTo(b) as -1 | 0 | 1;
}

/** Apply a percentage rate (GST, interest) and quantise. */
export function applyRate(base: Money, ratePercent: string | Decimal): Money {
  return quantise(base.times(new Decimal(ratePercent)).dividedBy(100));
}

/**
 * Split `total` across `weights` without losing or inventing a paisa.
 *
 * Naive proportional splitting silently drops the rounding remainder: ₹100 across three
 * flats gives 33.33 × 3 = ₹99.99 and a rupee vanishes. This distributes the remainder
 * one paisa at a time to the largest fractional losses first, so the parts always sum
 * exactly to the whole.
 *
 * That exactness is what makes a per-sq-ft maintenance run reconcile against the
 * society's bank statement — and a bill that does not reconcile in month one destroys
 * the committee's trust permanently.
 */
export function allocate(total: Money, weights: Decimal[]): Money[] {
  if (weights.length === 0) {
    throw new MoneyError("Cannot allocate across an empty set of weights.");
  }

  const weightTotal = weights.reduce((a, b) => a.plus(b), new Decimal(0));
  if (weightTotal.lessThanOrEqualTo(0)) {
    throw new MoneyError("Allocation weights must sum to a positive value.");
  }

  const raw = weights.map((w) => new Decimal(total).times(w).dividedBy(weightTotal));
  const parts = raw.map((r) => quantise(r as Money));

  const target = quantise(total);
  const allocated = parts.reduce((a, b) => a.plus(b), new Decimal(0));
  let remainder = target.minus(allocated);

  if (remainder.isZero()) return parts;

  const step = remainder.isPositive() ? new Decimal("0.01") : new Decimal("-0.01");
  // Largest fractional loss corrected first — the conventional, defensible order.
  const order = raw
    .map((r, i) => ({ i, loss: r.minus(parts[i]!) }))
    .sort((a, b) =>
      remainder.isPositive()
        ? b.loss.comparedTo(a.loss)
        : a.loss.comparedTo(b.loss),
    );

  const steps = remainder.abs().dividedBy("0.01").toNumber();
  for (let n = 0; n < steps; n++) {
    const idx = order[n % order.length]!.i;
    parts[idx] = parts[idx]!.plus(step) as Money;
  }

  return parts;
}

/**
 * Convert to paise at the payment-gateway boundary only.
 *
 * Razorpay works in integer paise. Converting here rather than earlier keeps every
 * internal calculation in Decimal, so the amount charged always equals the amount
 * invoiced.
 */
export function toPaise(amount: Money): number {
  return quantise(amount).times(100).toDecimalPlaces(0).toNumber();
}

export function fromPaise(paise: number): Money {
  return money(new Decimal(paise).dividedBy(100));
}

/** Canonical string for storage and for `numeric(18,4)` binding. */
export function toDbString(amount: Money): string {
  return new Decimal(amount).toFixed(STORAGE_DP);
}

/** Indian-format display, e.g. `₹1,23,456.78`. */
export function format(amount: Money, withSymbol = true): string {
  const fixed = quantise(amount).toFixed(MONEY_DP);
  const [whole = "0", fraction = "00"] = fixed.replace("-", "").split(".");

  // Indian grouping: last three digits, then pairs. 1234567 → 12,34,567
  const last3 = whole.slice(-3);
  const rest = whole.slice(0, -3);
  const grouped = rest
    ? `${rest.replace(/\B(?=(\d{2})+(?!\d))/g, ",")},${last3}`
    : last3;

  const sign = new Decimal(amount).isNegative() ? "-" : "";
  return `${sign}${withSymbol ? "₹" : ""}${grouped}.${fraction}`;
}

File: apps/api/app/common/money.py
"""Money handling.

Two rules, both non-negotiable:

1. Money is `Decimal`, never `float`. Postgres columns are `numeric(18,4)` and asyncpg
   maps them to `Decimal` directly, so the boundary is clean in both directions.
2. Money is computed **server-side only**. Clients display what the API returns; a live
   total while a bill is being edited comes from `POST /billing/preview`. There is
   exactly one implementation of every billing rule, so the total a resident sees and
   the total filed for GST cannot differ by a paisa.

Rounding is half-up at 2 decimal places for anything presented or invoiced. Bankers'
rounding is wrong here — Indian statutory invoicing expects half-up, and a committee
comparing against a hand-built spreadsheet will treat a 1-paisa difference as a bug.
Intermediate values keep 4 decimal places; only the final amount is quantised.
"""

from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal, InvalidOperation
from typing import Final

CURRENCY: Final = "INR"

#: Storage precision. Matches numeric(18,4) in Postgres.
STORAGE_EXPONENT: Final = Decimal("0.0001")

#: Presentation and invoicing precision — rupees and paise.
MONEY_EXPONENT: Final = Decimal("0.01")

ZERO: Final = Decimal("0.00")


class MoneyError(ValueError):
    """Raised when a value cannot be safely interpreted as money."""


def money(value: str | int | Decimal) -> Decimal:
    """Build a money value.

    `float` is rejected deliberately — accepting it would reintroduce the binary
    floating-point rounding errors this module exists to prevent.
    """
    if isinstance(value, float):
        raise MoneyError(
            "float is not accepted for money. Pass a str, int or Decimal "
            "(e.g. money('1250.50'), not money(1250.50))."
        )
    try:
        return Decimal(value).quantize(STORAGE_EXPONENT, rounding=ROUND_HALF_UP)
    except InvalidOperation as exc:
        raise MoneyError(f"Not a valid money value: {value!r}") from exc


def quantise(value: Decimal) -> Decimal:
    """Round to rupees and paise, half-up. Use for any presented or invoiced amount."""
    return value.quantize(MONEY_EXPONENT, rounding=ROUND_HALF_UP)


def apply_rate(base: Decimal, rate_percent: Decimal) -> Decimal:
    """Apply a percentage rate (GST, interest) and quantise the result.

    >>> apply_rate(Decimal("7500.00"), Decimal("18"))
    Decimal('1350.00')
    """
    return quantise(base * rate_percent / Decimal("100"))


def allocate(total: Decimal, weights: list[Decimal]) -> list[Decimal]:
    """Split `total` across `weights` without losing or inventing a paisa.

    Naive proportional splitting leaves a remainder that silently vanishes. This
    distributes the rounding remainder one paisa at a time to the largest shares
    first, so the parts always sum exactly to the whole — which is what makes a
    per-sq-ft maintenance split reconcile against the society's bank statement.

    >>> sum(allocate(Decimal("100.00"), [Decimal(1), Decimal(1), Decimal(1)]))
    Decimal('100.00')
    """
    if not weights:
        raise MoneyError("Cannot allocate across an empty set of weights.")

    weight_total = sum(weights, Decimal(0))
    if weight_total <= 0:
        raise MoneyError("Allocation weights must sum to a positive value.")

    raw = [total * w / weight_total for w in weights]
    parts = [quantise(r) for r in raw]

    remainder = quantise(total) - sum(parts, Decimal(0))
    if remainder == 0:
        return parts

    step = Decimal("0.01") if remainder > 0 else Decimal("-0.01")
    # Largest fractional loss gets corrected first — the conventional, defensible order.
    order = sorted(range(len(parts)), key=lambda i: raw[i] - parts[i], reverse=remainder > 0)

    for i in order[: int(abs(remainder) / Decimal("0.01"))]:
        parts[i] += step

    return parts
